generate.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2021 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terramate
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/madlambda/spells/errutil"
+	"github.com/mineiros-io/terramate/hcl"
+)
+
+const (
+	// GeneratedTfFilename is the name of the terramate generated tf file.
+	GeneratedTfFilename = "_gen_terramate.tm.tf"
+
+	// GeneratedCodeHeader is the header added on all generated files.
+	GeneratedCodeHeader = "// GENERATED BY TERRAMATE: DO NOT EDIT\n\n"
+)
+
+// Generate will walk all the directories starting from basedir generating
+// code for any stack it finds as it goes along
+//
+// It will return an error if it finds any invalid terramate configuration files
+// of if it can't generate the files properly for some reason.
+//
+// The provided basedir must be an absolute path to a directory.
+func Generate(basedir string) error {
+	if !filepath.IsAbs(basedir) {
+		return fmt.Errorf("Generate(%q): basedir must be an absolute path", basedir)
+	}
+
+	info, err := os.Lstat(basedir)
+	if err != nil {
+		return fmt.Errorf("Generate(%q): checking basedir: %v", basedir, err)
+	}
+
+	if !info.IsDir() {
+		return fmt.Errorf("Generate(%q): basedir is not a directory", basedir)
+	}
+
+	stacks, err := ListStacks(basedir)
+	if err != nil {
+		return fmt.Errorf("Generate(%q): listing stack: %v", basedir, err)
+	}
+
+	var errs []error
+
+	for _, stack := range stacks {
+		// At the time the most intuitive way was to start from the stack
+		// and go up until reaching the basedir, looking for a config.
+		// Basically navigating from the order of precedence, since
+		// more specific configuration overrides base configuration.
+		// Not the most optimized way (re-parsing), we can improve later
+		tfcode, err := generateStackConfig(basedir, stack.Dir)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if tfcode == nil {
+			continue
+		}
+
+		genfile := filepath.Join(stack.Dir, GeneratedTfFilename)
+		errs = append(errs, os.WriteFile(genfile, tfcode, 0666))
+	}
+
+	if err := errutil.Chain(errs...); err != nil {
+		return fmt.Errorf("Generate(%q): %w", basedir, err)
```

Comment:
I ran on my test stacks (with an "after" block not supported in your branch) and I got:

```
$ terramate generate
2021/12/11 21:22:23 Generate("/home/i4k/src/mineiros/tf-stacks"): parsing config: malformed terramate block: invalid attribute "after": parsing config: malformed terramate block: invalid attribute "after"
```

It successfully generated the code for some stacks.

The error does not show the offending stack and file. I also think we can improve the message because the part `Generate(<path>)` is such internal/debug info. WDYT?

I'd expect an error message like:

```
2021/12/11 21:22:23 failed to generate code for stack "/home/i4k/src/mineiros/tf-stacks/stack-b": parsing config "terramate.tm.hcl": malformed terramate block: invalid attribute "after": parsing config: malformed terramate block: invalid attribute "after"
```

Reply:
Will do on a follow up PR